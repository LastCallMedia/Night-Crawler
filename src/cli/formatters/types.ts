// @flow

import Analysis from '../../analysis';

export interface ComparisonFormatter {
  format(reports: Array<Analysis>): string;
}
