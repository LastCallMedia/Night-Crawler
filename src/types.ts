export type CrawlerRequest = {
  url: string;
  driverOptions?: unknown;
  groups?: string[]
  [key: string]: unknown;
};

export type CrawlerResponse = {
  url: string;
  groups?: []
  error?: Error;
  [key: string]: unknown;
};

export type DriverResponse = {
  statusCode: number;
};

export interface Driver<ResponseType extends DriverResponse = DriverResponse> {
  /**
   * Fetch a single URL.
   *
   * The driver should return a promise which is only rejected in the case
   * where the response is a complete error.
   */
  fetch(req: CrawlerRequest): Promise<ResponseType>;

  /**
   * Collect data about a response.
   *
   * A driver may provide default data about the response, which is merged
   * into the collected data the crawler itself gathers.
   */
  collect(response: ResponseType): Record<string, unknown>;
}
