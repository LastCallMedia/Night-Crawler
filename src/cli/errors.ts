export class FailedAnalysisError extends Error {}
