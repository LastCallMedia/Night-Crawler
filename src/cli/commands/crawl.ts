import ora from 'ora';
import { EOL } from 'os';
import { promises as fsPromises } from 'fs';
import { FailedAnalysisError } from '../errors';
import formatConsole from '../formatters/console';
import formatJUnit from '../formatters/junit';
import { BuilderCallback } from 'yargs';
import { ConfigArgs } from '../index';
import { TestResult } from '../../testing/TestContext';
import { hasFailure } from '../util';

export interface CrawlCommandArgs extends ConfigArgs {
  concurrency?: number;
  silent?: boolean;
  json?: string;
  junit?: string;
  stdout?: NodeJS.WritableStream;
}

export const command = 'crawl [crawlerfile]';
export const describe =
  'crawls a defined set of URLs and runs tests against the received responses.';
export const builder: BuilderCallback<ConfigArgs, CrawlCommandArgs> = yargs => {
  yargs.option('concurrency', {
    alias: 'c',
    describe: 'number of requests allowed in-flight at once',
    type: 'number',
    required: true,
    default: 3
  });
  yargs.option('silent', {
    alias: 'n',
    describe: 'silence all output',
    type: 'boolean',
    default: false
  });
  yargs.option('json', {
    alias: 'j',
    describe: 'filename to write JSON report to',
    normalize: true,
    type: 'string',
    default: ''
  });
  yargs.option('junit', {
    alias: 'u',
    describe: 'filename to write JUnit report to',
    normalize: true,
    type: 'string',
    default: ''
  });
};
export const handler = async function(argv: CrawlCommandArgs): Promise<void> {
  const { crawler, tests, json = '', junit = '', concurrency = 3 } = argv;
  const stdout = argv.stdout ?? process.stdout;
  const spinner = ora({
    stream: stdout,
    prefixText: 'Crawling'
  }).start('Starting');

  const eachResults = new Map<string, TestResult>();
  const responses = [];

  for await (const response of crawler.crawl(concurrency)) {
    eachResults.set(response.url, tests.testResponse(response));
    responses.push(response);
    spinner.text = `Crawled ${responses.length}`;
  }
  const allResults = tests.testResponses(responses);
  spinner.succeed('Finished Crawling');

  stdout.write(formatConsole(eachResults, allResults) + EOL);

  if (json.length) {
    await fsPromises.writeFile(
      json,
      JSON.stringify({ each: eachResults, all: allResults }),
      'utf8'
    );
  }

  if (junit.length) {
    await fsPromises.writeFile(
      junit,
      formatJUnit(eachResults, allResults),
      'utf8'
    );
  }

  if (
    hasFailure(allResults) ||
    Array.from(eachResults.values()).some(hasFailure)
  ) {
    throw new FailedAnalysisError('Testing reported an error.');
  }
};
