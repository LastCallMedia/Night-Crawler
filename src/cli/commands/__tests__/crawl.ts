import { CrawlCommandArgs, handler } from '../crawl';
import stream from 'stream';
import os from 'os';
import fs from 'fs';
import Crawler from '../../../crawler';
import DummyDriver from '../../../driver/dummy';
import { FailedAnalysisError } from '../../errors';
import formatJUnit from '../../formatters/junit';
import yargs from 'yargs';

import * as crawlModule from '../crawl';
import TestContext from '../../../testing/TestContext';
import { CrawlerResponse } from '../../../types';

jest.mock('../../formatters/junit', () => {
  return jest.fn(() => 'THIS IS A JUNIT REPORT');
});
jest.mock('../../formatters/console', () => {
  return jest.fn((a, b, opts) => {
    return `CONSOLE_ANALYSIS:${opts.color ? 'color' : 'nocolor'}`;
  });
});

const mockedJUnit = (formatJUnit as unknown) as jest.Mock<typeof formatJUnit>;

function runWithHandler(
  argv: string,
  handler: (argv: CrawlCommandArgs) => void
): Promise<Record<string, unknown>> {
  let invoked = 0;
  const cmd = Object.assign({}, crawlModule, {
    handler: (argv: CrawlCommandArgs) => {
      invoked++;
      handler(argv);
    }
  });

  return new Promise((res, rej) => {
    yargs
      .command(cmd)
      .parse(argv, (err: Error, argv: Record<string, unknown>) => {
        if (err) return rej(err);
        if (!invoked) return rej(new Error('handler was not invoked'));
        res(argv);
      });
  });
}
describe('Crawl Command', function() {
  it('Has defaults', function() {
    return runWithHandler('crawl', argv => {
      expect(argv.silent).toEqual(false);
      expect(argv.json).toEqual('');
      expect(argv.junit).toEqual('');
      expect(argv.concurrency).toEqual(3);
    });
  });
  it('Passes silent', function() {
    return runWithHandler('crawl --silent', argv => {
      expect(argv.silent).toEqual(true);
    });
  });
  it('Passes junit', function() {
    return runWithHandler('crawl --junit foo/bar.xml', argv => {
      expect(argv.junit).toEqual('foo/bar.xml');
    });
  });
  it('Passes json', function() {
    return runWithHandler('crawl --json foo/bar.json', argv => {
      expect(argv.json).toEqual('foo/bar.json');
    });
  });
  it('Passes concurrency', function() {
    return runWithHandler('crawl --concurrency 5', argv => {
      expect(argv.concurrency).toBe(5);
    });
  });
});

describe('Crawl Handler', function() {
  let stdout: stream.PassThrough;
  beforeEach(function() {
    stdout = new stream.PassThrough();
  });

  it('Executes the crawl', async function() {
    // eslint-disable-next-line
    const crawl = jest.fn(async function*() {});
    const crawler = new Crawler([]);
    const tests = new TestContext();
    crawler.crawl = crawl;
    await handler({ stdout, crawler, tests, concurrency: 1 });
    expect(crawl).toHaveBeenCalledWith(1);
  });

  it('Displays output to indicate the status of the crawl', async function() {
    const crawler = new Crawler(
      [{ url: 'http://example.com' }],
      new DummyDriver()
    );
    await handler({ stdout, crawler, tests: new TestContext() });
    const output = stdout.read().toString();
    expect(output).toMatchSnapshot();
  });

  it('Outputs analysis at the end of the crawl if the output is not silent', async function() {
    const crawler = new Crawler([], new DummyDriver());
    const tests = new TestContext();
    await handler({ stdout, crawler, tests });
    expect(stdout.read().toString()).toContain('CONSOLE_ANALYSIS:color');
  });

  it('Throws an error if the analysis contains failures', async function() {
    const crawler = new Crawler(
      [{ url: 'https://example.com' }],
      new DummyDriver()
    );
    const tests = new TestContext();
    tests.each('testing', () => {
      throw new Error('test');
    });
    await expect(handler({ stdout, crawler, tests })).rejects.toBeInstanceOf(
      FailedAnalysisError
    );
  });

  it('Should stop the crawl if setup fails', async function() {
    const crawler = new Crawler(
      // eslint-disable-next-line require-yield
      (async function*(): AsyncIterable<CrawlerResponse> {
        throw new Error('Oh no!');
      })(),
      new DummyDriver()
    );
    const p = handler({
      stdout,
      crawler,
      tests: new TestContext()
    });
    return expect(p).rejects.toThrow('Oh no!');
  });

  it('Should save a junit report if requested', async function() {
    const filename = `${os.tmpdir()}/nightcrawler-${Math.floor(
      Math.random() * 10000
    )}`;
    const crawler = new Crawler([], new DummyDriver());

    await handler({
      junit: filename,
      stdout,
      crawler,
      tests: new TestContext()
    });
    expect(mockedJUnit).toHaveBeenCalledWith(new Map(), new Map());
    await expect(
      fs.promises.readFile(filename, { encoding: 'utf-8' })
    ).resolves.toEqual('THIS IS A JUNIT REPORT');
    await fs.promises.unlink(filename);
  });

  it('Should save a valid JSON run report if requested', async function() {
    const filename = `${os.tmpdir()}/nightcrawler-${Math.floor(
      Math.random() * 10000
    )}`;
    const crawler = new Crawler([], new DummyDriver());
    await handler({
      json: filename,
      stdout,
      crawler,
      tests: new TestContext()
    });
    await expect(fs.promises.stat(filename)).resolves.toBeTruthy();
    await fs.promises.unlink(filename);
  });
});
