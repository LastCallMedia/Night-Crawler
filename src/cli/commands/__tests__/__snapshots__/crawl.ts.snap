// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crawl Handler Displays output to indicate the status of the crawl 1`] = `
"- Starting
Crawling [32mâœ”[39m Finished Crawling
CONSOLE_ANALYSIS:color
"
`;
